sync/group.go

```diff
@@ -0,0 +1,99 @@
+package sync
+
+import (
+	"context"
+	"sync"
+)
+
+// A Group is a collection of goroutines working on subtasks that are part of
+// the same overall task.
+//
+// Use NewGroup to create a new group.
+type Group struct {
+	ctx     context.Context
+	cancel  context.CancelCauseFunc
+	wg      sync.WaitGroup
+	sem     chan struct{}
+	errOnce sync.Once
+	err     error
+}
+
+// NewGroup returns a new group and an associated Context derived from ctx.
+//
+// The derived Context is canceled the first time a function passed to Go
+// returns a non-nil error or the first time Wait returns, whichever occurs
+// first.
+//
+// limit < 1 means no limit on the number of active goroutines.
+func NewGroup(ctx context.Context, limit int) (*Group, context.Context) {
+	ctx, cancel := context.WithCancelCause(ctx)
+	g := &Group{
+		ctx:    ctx,
+		cancel: cancel,
+	}
+	if limit > 0 {
+		g.sem = make(chan struct{}, limit)
+	}
+	return g, ctx
+}
+
+// Go calls the given function in a new goroutine.
+// It blocks until the new goroutine can be added without the number of
+// active goroutines in the group exceeding the configured limit.
+//
+// The first call to return a non-nil error cancels the group's context.
+// The error will be returned by Wait.
+//
+// If the group was created by calling NewGroup with limit < 1, there is no
+// limit on the number of active goroutines.
+//
+// If the group's context is canceled, routines that have not executed yet due to the limit won't be executed.
+// Additionally, there is a best effort not to execute `f()` once the context is canceled
+// and that happens whether or not a limit has been specified.
+func (g *Group) Go(f func() error) {
+	if err := g.ctx.Err(); err != nil {
+		g.errOnce.Do(func() {
+			g.err = g.ctx.Err()
+			g.cancel(g.err)
+		})
+		return
+	}
```

Comment:
Why do we need this block? :thinking:

Reply:
Just checking for context before waiting on group limits.
If the group didn't have a limit, we weren't checking for context before `go func() {...}()`.

Reply:
But it's OK because we don't define a limit so we're OK with creating a new go routine. Once that happens inside the go routine we do:

```go
	g.wg.Add(1)
	go func() {
		err := g.ctx.Err()
		if err == nil {
			err = f()
		}
```

So `f()` won't get executed anyway. I think we can simplify and remove that block, it complicates things unnecessarily imho.

If you remove it all tests pass but one i.e. `TestNoInitGroup`, probably because you don't have a `Wait()` there.

Reply:
The motivation was to not even start the goroutine.